src/Domain/Stager.php

```diff
@@ -84,12 +108,28 @@ private function validatePreconditions(): void
         }
     }
 
-    private function runCommand(): void
+    /**
+     * @throws \Symfony\Component\Process\Exception\RuntimeException
+     */
+    private function runCommand(): string
     {
-        $this->processFactory
+        $process = $this->processFactory
             ->create(array_merge([
                 'composer',
-                sprintf('--working-dir=%s', $this->stagingDir),
-            ], $this->command));
+                "--working-dir={$this->stagingDir}",
+            ], $this->composerCommand));
+        try {
+            $process->mustRun();
+            return $this->parseOutput($process);
+        } catch (\Symfony\Component\Process\Exception\ProcessFailedException $e) {
+            throw new ProcessFailedException($e->getMessage());
+        }
+    }
+
+    private function parseOutput(Process $process): string
+    {
+        $output = [$process->getOutput(), $process->getErrorOutput()];
```

Comment:
I'm not sure I quite understand the purpose of this method. The fact that it puts error output after regular output also means that it might be tricky, in the event of an error, to figure out when during the execution of the staged command the error happened. This is, IMHO, another argument in favor of using Process's output streaming.

Reply:
Ah, you're right. That's ugly. I did it strictly for console output, which is none of this class's business--console output is the responsibility of the console command. I wonder what the best design alternative would be. 🤔 My first thought is that instead of returning a string to the caller I could return `$self` and provide `getOutput()` and `getErrorOutput()` methods that just pass through to the `$process`. That would preserve my layered architecture and give the caller greater flexibility. But I don't know what effect it would have on the ability to stream output. Since I haven't worked with that before, maybe you could tell me a little about it in a post scrum next week.

Reply:
I haven't used the output streaming much myself either; I think you can pass a callback/closure to Process::run() which is called repeatedly as the process runs and delivers output. The callback receives the latest output, from both the standard output and error pipes, and can do with it what it pleases.

Reply:
> IMHO, we should use the ExecutableFinder and allow power users to specify a path to a binary. 

For the case where composer-stager is run from the command line, yes, that sounds good.

We can punt this to a followup, but just commenting here so it's written down somewhere: for the Drupal use case, I'm hoping we can run this without executing separate processes. Drupal will have a dependency both on composer/composer and on php-tuf/composer-stager. In such a case, it would be good for Drupal to be able to invoke the composer-stager commands via a PHP API (not via a process executer) and to perhaps pass to it a $composer object (or some other object from composer/composer) that composer-stager could use to invoke commands on. What do you think: does that seem sensible, and is there a way to do it without making composer-stager have a production dependency on composer/composer? I suppose it would need to have a dependency on some of the interfaces from there though?

Reply:
The `parseOutput()` method has been eliminated in favor of a callback as described in my reply to another related comment, enabling output streaming.

@effulgentsia, it is already the case in the current design that Drupal could invoke `composer-stager` in-process (via `Stager::stage()`). But as you point out, _that_ spawns a separate `composer` process. And you are correct that a dependency on Composer internals _would_ require `composer/composer` to be present in production. It would be _possible_ (though by no means easy) to look for `composer/composer` on the host and use _it_ rather than having a strict _Composer_ dependency (thus decoupling Drupal from it). Of course, that would have its challenges, such as the possibility that the host is using a Composer Phar, and require fallback behavior. I'll resolve this conversation for now and just let it stand as documentation in case we choose to pursue such an approach in the future.